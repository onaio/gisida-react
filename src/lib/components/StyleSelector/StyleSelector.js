import React, { Component } from 'react';
import { connect } from 'react-redux'
import PropTypes from 'prop-types';
import { Actions } from 'gisida';
import './StyleSelector.scss';


const mapStateToProps = (state, ownProps) => {
  const MAP = state[ownProps.mapId] || {}
  return {
    MAP,
    styles: state.STYLES,
    showFilterPanel: MAP.showFilterPanel
  }
}

export class StyleSelector extends Component {

  changeStyle = (e) => {
    const style = e.target.value;
    this.props.dispatch(Actions.changeStyle(this.props.mapId, style));
  }

  render() {
    const { styles, mapId } = this.props;
    const sidebarOffset = '0';
    
    return (
      <div className="leaflet-left leaflet-top leaflet-right layer-selector" style={{ right: sidebarOffset }}>
        <div aria-haspopup="true" className="leaflet-control leaflet-control-layers">
          <a title="styles" className="leaflet-control-layers-toggle"> </a>
          <form className="leaflet-control-layers-list">
            <div className="leaflet-control-layers-base">
              {(styles && styles.length > 0) ?
                styles.map(s =>
                  (<label key={`label_${s.label}-${this.props.mapId}`} htmlFor={`${s.label}-${this.props.mapId}`}>
                    <input  
                      readOnly
                      id={`${s.label}-${this.props.mapId}`}
                      key={`input_${s.label}-${this.props.mapId}`}
                      type="radio"
                      name="leaflet-base-layers"
                      className="leaflet-control-layers-selector"
                      value={s.url}
                      onClick={e => this.changeStyle(e)}
                      checked={(!!s[mapId] && !!s[mapId].current) || !!s.current }
                    />
                    <span>{s.label}</span>
                  </label>)) : <span>No available styles</span>
              }
            </div>
            <div className="leaflet-control-layers-overlays" />
          </form>
        </div>
      </div>)
  }
}

StyleSelector.propTypes = {
  styles: PropTypes.arrayOf(PropTypes.any).isRequired,
  showFilterPanel: PropTypes.bool
};

export default connect(mapStateToProps)(StyleSelector);
