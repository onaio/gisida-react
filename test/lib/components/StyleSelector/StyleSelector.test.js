import React from 'react';
import StyleSelector from '../../../../src/lib/components/StyleSelector/StyleSelector'
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { Actions } from 'gisida';

const STYLES = [
  {
    label: 'Satelitte',
    url: 'mapbox://styles/mapbox/satellite-v9',
  }
]
const initialState = {
  'map-1': { showFilterPanel: true },
  STYLES,
}
const mockStore = configureStore()
const store = mockStore(initialState);

describe('StyleSelector', () => {

  it('renders without crashing', () => {
    shallow(
      <StyleSelector 
        mapId='map-1'
        store={store}
        styles={[]}
      />
    );
  });

  it('should set props', () => {
    const wrapper = mount(
      <StyleSelector 
        mapId='map-1'
        store={store}
        styles={[]}
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();

    // Add a style
    STYLES.push({
      label: 'Satelitte Streets',
      url: 'mapbox://styles/mapbox/satellite-streets-v9',
    },)
		wrapper.setProps({STYLES });
    expect(wrapper.props().STYLES).toEqual(STYLES);
    
    // change style
    const changeStyleAction = jest.spyOn(Actions, 'changeStyle');
    wrapper.find(`#${STYLES[0].label}-map-1`).simulate('change');
    expect(changeStyleAction).toHaveBeenCalledWith('map-1', STYLES[0].url);
  })
});
