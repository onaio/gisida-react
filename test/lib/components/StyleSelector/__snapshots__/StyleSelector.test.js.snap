// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleSelector should set props 1`] = `
<Connect(StyleSelector)
  mapId="map-1"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  styles={Array []}
>
  <StyleSelector
    MAP={
      Object {
        "showFilterPanel": true,
      }
    }
    dispatch={[Function]}
    mapId="map-1"
    showFilterPanel={true}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    styles={
      Array [
        Object {
          "label": "Satelitte",
          "url": "mapbox://styles/mapbox/satellite-v9",
        },
      ]
    }
  >
    <div
      className="leaflet-left leaflet-top leaflet-right layer-selector"
      data-html2canvas-ignore={true}
      style={
        Object {
          "right": "0",
        }
      }
    >
      <div
        aria-haspopup="true"
        className="leaflet-control leaflet-control-layers"
      >
        <a
          className="leaflet-control-layers-toggle"
          title="styles"
        >
           
        </a>
        <form
          className="leaflet-control-layers-list"
        >
          <div
            className="leaflet-control-layers-base"
          >
            <label
              htmlFor="Satelitte-map-1"
              key="label_Satelitte-map-1"
            >
              <input
                className="leaflet-control-layers-selector"
                defaultValue="mapbox://styles/mapbox/satellite-v9"
                id="Satelitte-map-1"
                key="input_Satelitte-map-1"
                name="leaflet-base-layers"
                onChange={[Function]}
                readOnly={true}
                type="radio"
              />
              <span>
                Satelitte
              </span>
            </label>
          </div>
          <div
            className="leaflet-control-layers-overlays"
          />
        </form>
      </div>
    </div>
  </StyleSelector>
</Connect(StyleSelector)>
`;
